File: leetcode2025/2025_01/leetcode2025-01-05.py
from typing import List

class ATM:

    def __init__(self):
        self.bank = [0]*5

    def deposit(self, banknotesCount: List[int]) -> None:
        for i in range(5):
            self.bank[i] += banknotesCount[5-i-1]

    def withdraw(self, amount: int) -> List[int]:
        res = [0]*5
        money = [500,200,100,50,20]
        tot = 0
        for i in range(5):
            if self.bank[i]*money[i] <= amount:
                amount -= self.bank[i]*money[i]
                res[i] = self.bank[i]
            else:
                res[i] += amount//money[i]
                amount -= res[i]*money[i]
        if amount == 0:
            for i in range(5):
                self.bank[i] -= res[i]
            return res[::-1]
        return [-1]


# Your ATM object will be instantiated and called as such:
# obj = ATM()
# obj.deposit(banknotesCount)
# param_2 = obj.withdraw(amount)

if __name__ == "__main__":
    atm = ATM()
    op = ['d','w','d','w','w']
    d = [[0,0,1,2,1],[0,1,0,1,1]]
    w = [600,600,550]
    di,wi = 0,0
    for s in op:
        if s == 'd':
            atm.deposit(d[di])
            di += 1
        else:
            print(atm.withdraw(w[wi]))
            wi += 1

File: leetcode2025/2025_01/leetcode2025-01-05.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

class ATM {
private:
    vector<int> bank;
public:
    ATM() {
        bank.resize(5, 0);
    }
    
    void deposit(vector<int> banknotesCount) {
        for (int i=0; i<5; i++){
            bank[i] += banknotesCount[i];
        }
    }
    
    vector<int> withdraw(int amount) {
        vector<int> res(5, 0);
        vector<int> money = {20,50,100,200,500};
        int tot = 0;
        for (int i=4; i>=0; i--){
            if (1ll*bank[i]*money[i] <= 1ll*amount){
                amount -= 1ll*bank[i]*money[i];
                res[i] = bank[i];
            }
            else{
                res[i] += amount/money[i];
                amount -= 1ll*res[i]*money[i];
            }
        }
        if (amount == 0){
            for (int i=0; i<5; i++){
                bank[i] -= res[i];
            }
            return res;
        }
        return {-1};
    }
};

/**
 * Your ATM object will be instantiated and called as such:
 * ATM* obj = new ATM();
 * obj->deposit(banknotesCount);
 * vector<int> param_2 = obj->withdraw(amount);
 */

int main(){
    ATM atm;
    vector<string> op = {"d", "w", "d", "w", "w"};
    vector<vector<int>> d = {{0,0,1,2,1}, {0,1,0,1,1}};
    vector<int> w = {600, 600, 550};
    int di = 0, wi = 0;
    for (auto s : op){
        if (s == "d"){
            atm.deposit(d[di]);
            di++;
        }
        else{
            vector<int> res = atm.withdraw(w[wi]);
            wi++;
            for (auto i : res){
                cout << i << " ";
            }
            cout << endl;
        }
    }
}
